feat(nx-monorepo): nx composition and caching improvements

JeremyJonas opened the pull request:
- fix(nx-monorepo): resolve postinstall for non-node project
- refactor(nx-monorepo): use public project.subprojects
- chore: upgrade project to ^0.71.48
- feat(nx-monorepo): infer nx input/output by project type
  - enabled only for pdk repo, but can be enabled by consumers as well as is still experimental
- refactor(nx-monorepo): make nx configs composible
- feat(nx-monorepo): make NX_NON_DEFAULT_HASHER opt-in
  - Now we use the default nx hasher by default

JeremyJonas committed:
feat(nx-monorepo): make NX_NON_DEFAULT_HASHER opt-in

JeremyJonas committed:
refactor(nx-monorepo): make nx configs composible

Refactor nx config stuff into NxWorkspace and NxProject components
as composible pattern, including removing propagation of options.

JeremyJonas committed:
feat(nx-monorepo): infer nx input/output by project type

- opt-in feature that is enabled for pdk repo

JeremyJonas committed:
chore: upgrade project to ^0.71.49

Adds `project.subprojects` as public

JeremyJonas committed:
fix(nx-monorepo): resolve postinstall for non-node project

JeremyJonas committed:
refactor(nx-monorepo): use public project.subprojects

Remove temp hack to provide access to subprojects now that
projen has made this public

JeremyJonas commented on packages/nx-monorepo/src/components/nx-project/index.ts:
```diff
@@ -0,0 +1,277 @@
+];
+
+/**
+ * Component which manged the project specific NX Config and is added to all NXMonorepo subprojects.
```
*manages

JeremyJonas commented on packages/nx-monorepo/src/components/nx-project/index.ts:
```diff
@@ -0,0 +1,277 @@
+   * @see https://nx.dev/reference/project-configuration#implicitdependencies
+   */
+  public implicitDependencies: string[] = [];
+  /**
+   * Explicit list of scripts for Nx to include.
+   * @see https://nx.dev/reference/project-configuration#ignoring-package.json-scripts
+   */
+  public includedScripts: string[] = [];
```
Need to implement this in synth below - forgot to do this

JeremyJonas commented on packages/nx-monorepo/src/components/nx-project/index.ts:
```diff
@@ -0,0 +1,277 @@
+    super(project);
+
+    this.file =
+      (project.tryFindObjectFile("project.json") as JsonFile) ||
```
If "project.json" already exists, then the below `obj` will not get synthesized - need to fix this

JeremyJonas commented on packages/nx-monorepo/src/components/nx-project/targets.ts:
```diff
@@ -0,0 +1,109 @@
+  readonly dependsOn?: boolean | string[];
+}
+
+export function inferBuildTarget(
```
docs

JeremyJonas commented on packages/nx-monorepo/src/components/nx-workspace.ts:
```diff
@@ -0,0 +1,243 @@
+import { Nx } from "../nx-types";
+import { asUndefinedIfEmpty, deepMerge } from "../utils/common";
+
+export const ALWAYS_IGNORE: string[] = [".tmp", ".env", ".pytest_cache"];
```
docs

JeremyJonas commented on packages/nx-monorepo/src/components/nx-workspace.ts:
```diff
@@ -0,0 +1,243 @@
+];
+
+/**
+ * Component which manged the workspace specific NX Config for the root monorepo.
```
*manages

JeremyJonas commented on packages/nx-monorepo/src/nx-monorepo.ts:
```diff
@@ -89,10 +89,10 @@ export interface MonorepoUpgradeDepsOptions {
+  // /**
+  //  * Configuration for NX.
+  //  */
+  // readonly nxConfig?: Nx.WorkspaceConfig;
```
Remove this

JeremyJonas commented on packages/nx-monorepo/src/nx-monorepo.ts:
```diff
@@ -667,9 +549,30 @@ export class NxMonorepoProject extends TypeScriptProject {
   }
 
+  /**
+   * Ensures that all non-root projects have NxProjectConfig applied.
```
*NxProject

JeremyJonas commented on packages/nx-monorepo/src/nx-monorepo.ts:
```diff
@@ -667,9 +549,30 @@ export class NxMonorepoProject extends TypeScriptProject {
   preSynthesize(): void {
     NodePackageUtils.removeProjenScript(this);
 
+    // Calling being "super" to ensure proper preSynth of NxProject component and its nested components
```
*before

JeremyJonas commented on packages/nx-monorepo/src/nx-types.ts:
```diff
@@ -1,37 +1,60 @@
    * Configuration for nx targetDependencies.
    */
-  export type TargetDependencies = { [target: string]: TargetDependency[] };
+  export type ITargetDependencies = { [target: string]: ITargetDependency[] };
```
From "I" prefix, just a type

JeremyJonas commented on packages/nx-monorepo/src/nx-types.ts:
```diff
@@ -1,37 +1,60 @@
    * Configuration for project specific targets.
    */
-  export type ProjectTargets = { [target: string]: ProjectTarget };
+  export type IProjectTargets = { [target: string]: IProjectTarget };
```
Remove "I" prefix, just a type

JeremyJonas committed:
fix(nx-monorepo): finalize NxProject impl and cleanup

JeremyJonas committed:
fix(jsii): use type mapping for jsii ignore patch

Using type instead of interface results in a @jsii.map with
elementType as expected for {[key:string]: xxx} types.

agdimech commented on packages/cdk-graph/src/config/index.ts:
```diff
@@ -13,7 +13,7 @@ export const CDK_OUTDIR_VAR = "<cdk.out>";
 
 /** CdkGraph configuration definition */
-export interface CdkGraphConfig {
+export type CdkGraphConfig = {
```
Do types means index types will work cross language?

JeremyJonas replied:
Using `type` gives the following JSII, which is `map` - so looks like should work right?

```
        {
          "abstract": true,
          "docs": {
            "see": "https://nx.dev/reference/project-configuration",
            "stability": "stable",
            "summary": "Targets configuration."
          },
          "immutable": true,
          "locationInModule": {
            "filename": "src/nx-types.ts",
            "line": 246
          },
          "name": "targets",
          "optional": true,
          "type": {
            "collection": {
              "elementtype": {
                "fqn": "@aws-prototyping-sdk/nx-monorepo.Nx.IProjectTarget"
              },
              "kind": "map"
            }
          }
        }
      ],
```

agdimech commented on packages/nx-monorepo/src/components/nx-project/index.ts:
```diff
@@ -0,0 +1,304 @@
+        const _target = this.targets[task.name] || {};
+        _target.executor = _target.executor || "nx:run-commands";
+        _target.options = {
+          command: NodePackageUtils.command.downloadExec(
```
why do we need download exec here? I would assume projen is already installed so we should just need to call exec?

JeremyJonas replied:
I copied it from the code you wrote before 😊 ... assuming was to do with non-node based package/repo support? Since before was only used for non-node projects.

JeremyJonas replied:
I have removed the `dlx` stuff - using `NodePackageUtils.command.projen(packageManager, task.name)` now

agdimech commented on packages/nx-monorepo/src/utils/common.ts:
```diff
@@ -0,0 +1,123 @@
+ *
+ * `undefined`s will cause a value to be deleted if destructive is enabled.
+ */
+export function deepMerge(
```
Is the different to the deepMerge in projen/lib?

JeremyJonas replied:
Yes, it adds functionality to "append to array"... I will update description to make this clear

agdimech commented on packages/nx-monorepo/src/utils/node.ts:
```diff
@@ -19,6 +24,14 @@ export namespace NodePackageUtils {
 
+  /** Indicates if project is a node based project */
+  export function isNodeProject(project: Project): boolean {
+    if (project instanceof NodeProject) {
```
Can be simplified to:

`return project instanceof NodeProject || tryFindNodePackage(project, false) != null;`

agdimech commented on private/pdk-project.ts:
```diff
@@ -182,86 +183,50 @@ export abstract class PDKProject extends JsiiProject {
-
-    this.nx = nx;
-  }
+  // /**
```
Any reason this just can't be deleted?

JeremyJonas replied:
Oh it should have been, not sure how missed that

agdimech commented on private/pdk-project.ts:
```diff
@@ -345,16 +310,12 @@ class PDKRelease extends Release {
- */
-export function cloneDeep(value: any): any {
-  return JSON.parse(JSON.stringify(value));
+    NxProject.of(project)?.addBuildTargetFiles(
```
Can't this just be:

`NxProject.of(project)?.addBuildTargetFiles(
      [],
      ["{projectRoot}/LICENSE_THIRD_PARTY"],
      true
    );`

JeremyJonas replied:
That will generate inputs of `["!{projectRoot}/LICENSE_THIRD_PARTY", "!{projectRoot}/LICENSE_THIRD_PARTY/**/*"]`, because we don't know if it is a file or directory. Since we know this is just a file it is best to just be explicit.

NX input seems to require the `/**/*` for directories.

agdimech replied:
How about:

`NxProject.of(project)?.addBuildTargetFiles( [], [File.of("{projectRoot}/LICENSE_THIRD_PARTY")], true );`

agdimech commented on private/projects/pdk-monorepo-project.ts:
```diff
@@ -90,33 +54,26 @@ export class PDKMonorepoProject extends NxMonorepoProject {
     });
 
+    // Turn on automatic target inference for this repo
+    this.nx._autoInferProjectTargets = true;
```
Is there a nicer way to set this?

JeremyJonas replied:
What you have in mind? it is experimental and I expect we will enable it by default once we have more robust coverage and feedback - at which time was planning to change, which I why I am ok with an ugly implementation which stands out a bit.

agdimech replied:
`this.nx.autoInferProjectTargets()`

agdimech commented on private/projects/pdk-monorepo-project.ts:
```diff
@@ -90,33 +54,26 @@ export class PDKMonorepoProject extends NxMonorepoProject {
+    this.nx.useNxCloud(
+      "OWJmZDJmZmEtNzk5MC00OGJkLTg3YjUtNmNkZDk1MmYxZDZkfHJlYWQ="
+    );
+    this.nx.cacheableOperations.push("generate");
```
Can we also set the cacheDirectory to `.nx/cache`?

JeremyJonas replied:
Ah yes, forgot about that

agdimech commented on packages/nx-monorepo/src/nx-monorepo.ts:
```diff
@@ -363,70 +316,15 @@ export class NxMonorepoProject extends TypeScriptProject {
-    !NxProject.of(subproject) &&
-      subproject._addComponent(new NxProject(subproject));
-    super._addSubProject(subproject);
+    this.nx = NxWorkspace.of(this) || new NxWorkspace(this);
```
Can we also have an ensure method in NxWorkspace?

JeremyJonas replied:
NxWorkspace should only be set once by the root, and is not required. While do to construct/synth order of operations, the NxProject might not have been instantiated when it is needed - so `ensure` makes sense. I don't think we want the same loose functionality around NxWorkspace.

But if have a use case for it then yes is easy to add

agdimech replied:
Makes sense, please ignore.

agdimech reviewed (commented):
Really great improvements mate!

JeremyJonas committed:
fix(nx-monorepo): nx improvements

JeremyJonas committed:
fix(nx-monorepo): postinstall, cacheDir, and workflow issues

- disable the postinstall for non-node packages until can resolve issues
- remove non default hasher in workflows
- disable cacheDirectory and add experimental warning

agdimech reviewed (approved):
Great work!
